File: ios/player/HybridNitroPlayerPreview.swift
//
//  HybridNitroPlayerPreview.swift
//  NitroPlay
//

import AVFoundation
import Foundation

extension HybridNitroPlayer {
  // MARK: - First-frame / preview generation

  func cancelFirstFrameRequest() {
    firstFrameRequest?.cancel()
    firstFrameRequest = nil
    firstFrameTask?.cancel()
    firstFrameTask = nil
  }

  func cacheFirstFrameContext(sourceUri: String, width: Double, height: Double) {
    firstFrameContext = FirstFrameContext(
      sourceUri: sourceUri,
      headers: currentSourceConfig()?.headers,
      width: width,
      height: height
    )
  }

  func emitFirstFrame(uri: String, width: Double, height: Double, sourceUri: String, fromCache: Bool) {
    let data = onFirstFrameData(
      uri: uri,
      width: width,
      height: height,
      sourceUri: sourceUri,
      fromCache: fromCache
    )
    firstFrame = data
    _eventEmitter?.onFirstFrame(data)
  }

  func markReadyToDisplay() {
    guard !isReleased else { return }
    readyToDisplay = true
    requestFirstFrameIfNeeded()
    emitPlaybackState()
  }

  func requestFirstFrameIfNeeded() {
    guard !isReleased, hasActiveSource, readyToDisplay, firstFrame == nil else { return }
    guard let context = firstFrameContext else { return }
    let autoThumbnailEnabled = currentAutoThumbnailEnabled()

    switch currentPreviewMode() {
    case .manual:
      guard autoThumbnailEnabled else { return }
    case .listener:
      guard autoThumbnailEnabled || _eventEmitter?.hasOnFirstFrameListeners() == true else { return }
    case .always:
      break
    @unknown default:
      break
    }

    let generation = sourceGeneration
    if firstFrameTask != nil {
      return
    }

    guard let request = VideoPreviewRuntime.shared.startFirstFrameRequest(
      url: context.sourceUri,
      headers: context.headers,
      preview: currentSourceConfig()?.preview
    ) else {
      return
    }
    firstFrameRequest = request

    firstFrameTask = Task.detached(priority: .utility) { [weak self] in
      guard let self else { return }
      let result = await request.value()
      request.cancel()

      DispatchQueue.main.async { [weak self] in
        guard let self else { return }
        defer {
          if self.firstFrameRequest === request {
            self.firstFrameRequest = nil
            self.firstFrameTask = nil
          }
        }
        guard !self.isReleased, self.sourceGeneration == generation, self.hasActiveSource, self.readyToDisplay, self.firstFrame == nil else {
          return
        }
        guard let result else { return }
        self.emitFirstFrame(
          uri: result.uri,
          width: context.width,
          height: context.height,
          sourceUri: context.sourceUri,
          fromCache: result.fromCache
        )
      }
    }
  }

  private func currentPreviewMode() -> NitroSourcePreviewMode {
    currentSourceConfig()?.preview?.mode ?? .listener
  }

  private func currentAutoThumbnailEnabled() -> Bool {
    currentSourceConfig()?.preview?.autoThumbnail ?? true
  }
}

File: android/src/main/java/com/nitroplay/player/HybridNitroPlayerPreview.kt
package com.margelo.nitro.video

import android.os.Handler
import android.os.Looper
import com.nitroplay.hls.VideoPreviewRuntime

// First-frame / preview generation, split out of HybridNitroPlayer (mirror of the
// iOS HybridNitroPlayer preview extension). Extension functions keep the player's
// stored first-frame state in the host while grouping the generation logic.

internal fun HybridNitroPlayer.cancelFirstFrameRequest() {
  firstFrameRequest?.cancel()
  firstFrameRequest = null
  firstFrameTask?.cancel(true)
  firstFrameTask = null
  pendingFirstFrameGeneration = -1
}

internal fun HybridNitroPlayer.cacheFirstFrameContext(sourceUri: String, width: Double, height: Double) {
  firstFrameContext = HybridNitroPlayer.FirstFrameContext(
    sourceUri = sourceUri,
    headers = currentSourceConfig()?.headers,
    width = width,
    height = height
  )
}

internal fun HybridNitroPlayer.emitFirstFrame(uri: String, width: Double, height: Double, sourceUri: String, fromCache: Boolean) {
  val data = onFirstFrameData(
    uri = uri,
    width = width,
    height = height,
    sourceUri = sourceUri,
    fromCache = fromCache
  )
  firstFrame = data
  eventEmitter.onFirstFrame(data)
}

internal fun HybridNitroPlayer.requestFirstFrameIfNeeded() {
  if (isReleased || !hasActiveSource || !readyToDisplay || firstFrame != null) {
    return
  }
  val autoThumbnailEnabled = currentAutoThumbnailEnabled()

  when (currentPreviewMode()) {
    NitroSourcePreviewMode.MANUAL -> {
      if (!autoThumbnailEnabled) {
        return
      }
    }
    NitroSourcePreviewMode.LISTENER -> {
      if (!autoThumbnailEnabled && !eventEmitter.hasOnFirstFrameListeners()) {
        return
      }
    }
    NitroSourcePreviewMode.ALWAYS -> Unit
  }

  val context = firstFrameContext ?: return
  val generation = sourceGeneration
  if (pendingFirstFrameGeneration == generation) {
    return
  }
  pendingFirstFrameGeneration = generation
  val previewConfig = currentSourceConfig()?.preview
  val request = VideoPreviewRuntime.startFirstFrameRequest(
    context.sourceUri,
    context.headers,
    previewConfig
  ) ?: run {
    pendingFirstFrameGeneration = -1
    return
  }
  firstFrameRequest = request

  firstFrameTask = VideoPreviewRuntime.dispatchFirstFrameAwait {
    val preview = try {
      if (isReleased) {
        null
      } else {
        request.await()
      }
    } finally {
      request.cancel()
    }

    Handler(Looper.getMainLooper()).post {
      if (pendingFirstFrameGeneration == generation) {
        pendingFirstFrameGeneration = -1
        firstFrameTask = null
      }
      if (firstFrameRequest === request) {
        firstFrameRequest = null
      }

      if (
        isReleased ||
        sourceGeneration != generation ||
        !hasActiveSource ||
        !readyToDisplay ||
        firstFrame != null
      ) {
        return@post
      }

      preview?.let {
        emitFirstFrame(
          uri = it.uri,
          width = context.width,
          height = context.height,
          sourceUri = context.sourceUri,
          fromCache = it.fromCache
        )
      }
    }
  }
}

private fun HybridNitroPlayer.currentPreviewMode(): NitroSourcePreviewMode {
  return currentSourceConfig()?.preview?.mode ?: NitroSourcePreviewMode.LISTENER
}

private fun HybridNitroPlayer.currentAutoThumbnailEnabled(): Boolean {
  return currentSourceConfig()?.preview?.autoThumbnail ?: true
}
